### ISSUE TITLE
Standardized way to serialize/deserialize hanabi particle effects 

### ISSUE THREAD

**Issue by ethereumdegen** [djeedai/bevy_hanabi#270]
**Describe the solution you'd like**
There should be a standard way to serialize and deserialize these particle effects so they can exist as files. 

**Describe _why_ you want that solution. Is this related to a problem?**
With VFX as files, they can be more easily shared and loaded into a game or a game editor 

**Describe alternatives you've considered**
 

**Additional context**

**Comment by ethereumdegen** [comment]
I prototyped this here  https://github.com/ethereumdegen/bevy_hanabi_loader

**Comment by djeedai** [comment]
Yes this is a planned feature, although one pain point from the start has been that modifiers are trait objects, and trait object serialization is pretty hard (not supported by default in serde for example).

There's also a number of things that delayed implementation, because as soon as one version is out, people will want back-compat (otherwise there's no point having a serialized asset if you can't load it after upgrading).

- Bevy Asset V2 just landed, and was promising breaking changes. Serializing and deserializing of referenced assets like textures referenced by asset path was not easily doable before. Lack of import settings and `.meta` files also made things harder to handle.
- To some extent it's still unclear that the current strategy of using `typetag` is worth continuing; it's completely unrelated to Bevy's reflection and type system. There has been some talks about allowing serializing in Bevy with the `TypeRegistry` available as context but currently it's not possible.
- Some GPU resource management like textures and gradients is still very much hard-coded, and will disappear in the future. That would require some back-compat effort to handle.

So the current strategy is that I try to make most types reflected and serializable, without supporting an official asset format. If someone wants to define their own format they should be able to do so (like, just serialize the `EffectAsset` itself directly to _e.g._ RON). And if there's any blocker doing so, I'm very much open to bug fixes and suggestions to improve. And maybe with Asset V2 now out, this is the right time to start looking at this again; I just didn't find the time to do so yet, unfortunately.

**Comment by djeedai** [comment]
This issue is largely fixed, with the exception of some details in serialization:
- Dependency assets like texture paths are not yet handled.
- Effect hierarchies (parent/child) can't be expressed since `EffectAsset` describes a single effect.

Leaving that issue open for those, but aside from them in 0.19 there is a standard official Hanabi way to serialize assets via `EffectAsset::serialize()` and `EffectAsset::deserialize()`. This includes all modifiers, even user-defined ones. And this uses the Bevy type registry (`typetag` dependency was removed).
